### PR TITLE
Add support for CronJob type of components

### PR DESCRIPTION
**Pull request by victorv-sbp**
If you have a `CronJob` type of component the release system will not make the necessary changes in the `_patch_yaml_docs `method. Meaning that things like image and tag will not be replaced.

Added logic to also include these type of files (as they have a different structure than `Deployment`) and also refactored a bit the existing `_patch_*` methods.

### CODE REVIEW

File: devops/lib/component.py [hunk 1]
```diff
@@ -360,23 +360,16 @@ def _patch_generic(self, doc: dict):
...
-        self._patch_image_pull_secrets(doc)
-        self._patch_replicas(doc)
-
-    def _patch_stateful_set(self, doc: dict):
-        logger.info("Patching found StatefulSet")
-        self._patch_containers(doc)
-        self._patch_image_pull_secrets(doc)
-        self._patch_replicas(doc)
+    def _patch_cronjob(self, doc: dict):
+        spec = doc["spec"]["jobTemplate"]
```
Comment (joakimnordling) [inline devops/lib/component.py]: Should this variable be called `job_template` just to avoid confusion?

File: devops/tests/test_component.py [hunk 2]
```diff
@@ -78,3 +103,23 @@ def test_patch_replicas():
     c._patch_replicas(deploy)
 
     assert deploy["spec"]["replicas"] == 77
+
+
+def test_patch_cronjob():
+    cronjob = get_cronjob()
+    c = Component("service/test-cronjob")
+    c.replicas = 77
```
Comment (joakimnordling) [inline devops/tests/test_component.py]: I started wondering that what replicas would actually mean for a cron job. I couldn't find any reference to it at least on https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#writing-a-cron-job-spec. Could you @victorv-sbp please point me to some documentation about that?

Comment (victorv-sbp) [reply devops/tests/test_component.py]: Indeed, the replicas is nowhere to be found on the kubernetes documentation regarding Jobs.
We've supposed that replicas would just create more pods on job creation/trigger (as in deployment case) as the kubeval reported the spec with replicas as a valid one.

However, after more in dept research we've come to the conclusion that the replicas is not needed in a job definition.

We've changed the code accordingly, please review the changes when you get a chance.

Comment (joakimnordling) [reply devops/tests/test_component.py]: Looks good to me.

File: devops/tests/test_component.py [hunk 3]
```diff
@@ -24,11 +24,35 @@
...
+metadata:
+  name: test-cronjob
+spec:
+  schedule: '* * * * *'
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+            app: test-cronjob
```
Comment (soderluk) [inline devops/tests/test_component.py]: @victorv-sbp: I checked the kubernetes documentation for the CronJob kind, and found out that there is no `metadata` under the `template:` key in their example (https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#creating-a-cron-job). Is there a reason why this is defined here, or is it just by accident?

Comment (victorv-sbp) [reply devops/tests/test_component.py]: Indeed, it does not show in any of the official examples (probably in order to only display and underline the critical fields for a cronjob), but it is a valid yaml definition of a CronJob(kubeval, PyCharm k8s yaml validator and applying the definition on a cluster, all work with a `metadata` under the `template`). 
Also, if you apply a CronJob on a cluster without that metadata branch, you can later see that describing that CronJob will display a `metadata` field under the `template` area.

In conclusion the `metadata` area is not wrong but it's also not really important for a Job definition. It could be deleted without having any effect on the test, but considering that the deployment also has a metadata (with the same role and importance), I would keep it there just for consistency. If you still want to delete it then I will do so.

Comment (soderluk) [reply devops/tests/test_component.py]: I'd rather not have anything excessive if it's not needed nor used anyway. But I'm OK to have these in the tests, so that it's certain that the tests pass.

**Comment by joakimnordling** [comment]
I was actually still a bit wondering about the replicas for a CronJob, and especially why kubeval was not saying anything about that.

I just today realized that kubeval has a `--strict` flag that can be used to `Disallow additional properties not in schema`. Whit that flag and replicas set for the cronjob you'd get this output:
```
WARN - cronjob.yaml contains an invalid CronJob - replicas: Additional property replicas is not allowed
```

**Comment by lietu** [comment]
Should we then update the scripts to use that?

**Comment by joakimnordling** [comment]
Yes, I think it would make sense to add the `--strict` flag. Will submit a PR for that shortly.
